src/keysym.rs

```diff
@@ -20,66 +87,141 @@ pub enum Keysym {
+            KEYSYM_PAUSE => Ok(Pause),
+            KEYSYM_CAPS_LOCK => Ok(CapsLock),
+            KEYSYM_SUPER_LEFT => Ok(SuperLeft),
+            KEYSYM_SUPER_RIGHT => Ok(SuperRight),
+            KEYSYM_MENU => Ok(Menu),
+
+            KEYSYM_PAGE_DOWN => Ok(PageDown),
+            KEYSYM_LEFT => Ok(Left),
+            KEYSYM_UP => Ok(Up),
+            KEYSYM_RIGHT => Ok(Right),
+            KEYSYM_DOWN => Ok(Down),
+
+            f if (f >= KEYSYM_F1 && f <= KEYSYM_F12) => {
+                let n = f - KEYSYM_F1 + 1;
                 // TODO: handle cast
                 Ok(FunctionKey(n as u8))
             }
-            0xffe1 => Ok(ShiftLeft),
-            0xffe2 => Ok(ShiftRight),
-            0xffe3 => Ok(ControlLeft),
-            0xffe4 => Ok(ControlRight),
-            0xffe7 => Ok(MetaLeft),
-            0xffe8 => Ok(MetaRight),
-            0xffe9 => Ok(AltLeft),
-            0xffea => Ok(AltRight),
-
-            // TODO: figure out if there's a better way to map codes
-            other => {
-                let c = char::from_u32(other);
-                match c {
-                    // TODO: figure out what to do with these
-                    None => Ok(Unknown(other)),
-                    Some(v) => Ok(Utf32(v)),
-                }
-            }
+
+            KEYSYM_SHIFT_LEFT => Ok(ShiftLeft),
+            KEYSYM_SHIFT_RIGHT => Ok(ShiftRight),
+            KEYSYM_CTRL_LEFT => Ok(ControlLeft),
+            KEYSYM_CTRL_RIGHT => Ok(ControlRight),
+            KEYSYM_ALT_LEFT => Ok(AltLeft),
+            KEYSYM_ALT_RIGHT => Ok(AltRight),
+
+            KEYSYM_SCROLL_LOCK => Ok(ScrollLock),
+            KEYSYM_NUM_LOCK => Ok(NumLock),
+
+            KEYSYM_KP_ENTER => Ok(KeypadEnter),
+            KEYSYM_KP_SLASH => Ok(KeypadSlash),
+            KEYSYM_KP_ASTERISK => Ok(KeypadAsterisk),
+            KEYSYM_KP_MINUS => Ok(KeypadMinus),
+            KEYSYM_KP_PLUS => Ok(KeypadPlus),
+            KEYSYM_KP_7 => Ok(Keypad7),
+            KEYSYM_KP_HOME => Ok(KeypadHome),
+            KEYSYM_KP_8 => Ok(Keypad8),
+            KEYSYM_KP_UP => Ok(KeypadUp),
+            KEYSYM_KP_9 => Ok(Keypad9),
+            KEYSYM_KP_PGUP => Ok(KeypadPgUp),
+            KEYSYM_KP_4 => Ok(Keypad4),
+            KEYSYM_KP_LEFT => Ok(KeypadLeft),
+            KEYSYM_KP_5 => Ok(Keypad5),
+            KEYSYM_KP_EMPTY => Ok(KeypadEmpty),
+            KEYSYM_KP_6 => Ok(Keypad6),
+            KEYSYM_KP_RIGHT => Ok(KeypadRight),
+            KEYSYM_KP_1 => Ok(Keypad1),
+            KEYSYM_KP_END => Ok(KeypadEnd),
+            KEYSYM_KP_2 => Ok(Keypad2),
+            KEYSYM_KP_DOWN => Ok(KeypadDown),
+            KEYSYM_KP_3 => Ok(Keypad3),
+            KEYSYM_KP_PGDOWN => Ok(KeypadPgDown),
+            KEYSYM_KP_0 => Ok(Keypad0),
+            KEYSYM_KP_INSERT => Ok(KeypadInsert),
+            KEYSYM_KP_PERIOD => Ok(KeypadPeriod),
+            KEYSYM_KP_DELETE => Ok(KeypadDelete),
+
+            _ => Ok(Unknown(value)),
```

Comment:
Why is this a try-from if we unconditionally succeed?

Reply:
We chatted a bit about this offline, but -- yeah, you're right, this doesn't make a lot of sense. I had wanted to distinguish between keysyms that aren't mapped in this crate (because there are a lot of them) and invalid keysyms, hence the Unknown variant and making the conversion function a `try_from`. I used the variant a lot for debugging while wiring things up with the keyboard, but it's pretty useless in this interface for consumers. So I'll just convert the unknown case to an error.